zh_tw/news/_posts/2024-10-07-ruby-3-4-0-preview2-released.md

```diff
@@ -0,0 +1,136 @@
+---
+layout: news_post
+title: "Ruby 3.4.0 preview2 發布"
+author: "naruse"
+translator: "Bear Su"
+date: 2024-10-07 00:00:00 +0000
+lang: zh_tw
+---
+
+{% assign release = site.data.releases | where: "version", "3.4.0-preview2" | first %}
+我們很高興宣布 Ruby {{ release.version }} 發布了。
+
+## Prism
+
+預設的解析器從 parse.y 切換為 Prism。 [[Feature #20564]]
+
+## 語法變更
+
+* 沒有 `frozen_string_literal` 註解的檔案中的字串文字，現在行為會跟被凍結一樣。如果它們被修改則會發出棄用警告。
+  這些警告可以透過 `-W:deprecated` 或 `Warning[:deprecated] = true` 啟用。
+  若要停用此變更，您可以在執行 Ruby 時加上參數 `--disable-frozen-string-literal`。 [[Feature #20205]]
+
+* 新增 `it` 來引用 block 的傳入參數。 [[Feature #18980]]
+
+* 現在呼叫方法支援使用 Keyword splatting `nil`。
+  `**nil` 的處理方式與 `**{}` 類似，不會傳遞 keywords，也不會呼叫任何轉換方法。 [[Bug #20064]]
+
+* Block passing 不再允許作為索引。 [[Bug #19918]]
+
+* 不定長度參數不再允許作為索引。 [[Bug #20218]]
+
+## 核心類別更新
+
+注意：我們只列出特別的類別更新。
+
+* Exception
+
+  * Exception#set_backtrace 現在可接受 `Thread::Backtrace::Location` 陣列。
+    `Kernel#raise`、`Thread#raise` 和 `Fiber#raise` 也接受同樣的新格式。 [[Feature #13557]]
+
+* Range
+
+  * Range#size 如果 range 無法迭代的話，現在會拋出 TypeError。 [[Misc #18984]]
+
+## 相容性問題
+
+注意：不包含功能問題的修正。
+
+* 錯誤訊息與 backtrace 的顯示改變。
+  * 使用單引號 (`'`) 取代反引號 (`` ` ``) 作為起始引號。 [[Feature #16495]]
+  * 在方法名稱前顯示類別名稱 (僅當類別有固定名稱時)。 [[Feature #19117]]
+  * `Kernel#caller`、`Thread::Backtrace::Location` 的方法等等，也做了對應的改變。
+
+  ```
+  舊:
+  test.rb:1:in `foo': undefined method `time' for an instance of Integer
+          from test.rb:2:in `<main>'
```

Comment:
```suggestion
          from test.rb:2:in '<main>'
```

See https://github.com/ruby/www.ruby-lang.org/pull/3384

Reply:
#3384 just changed the New output. The Old output is not changed. I think we could keep it.

I run the command in the issue.

```ruby
docker run --platform linux/amd64 --rm -it ghcr.io/ruby/all-ruby env LANG=C.UTF-8 ALL_RUBY_SINCE=ruby-3.3 ./all-ruby -e 'def foo; time; end; foo'
```

The output:
```ruby
ruby-3.3.0          -e:1:in `foo': undefined local variable or method `time' for main (NameError)
                    
                    def foo; time; end; foo
                             ^^^^
                        from -e:1:in `<main>'
                exit 1
...
ruby-3.3.5          -e:1:in `foo': undefined local variable or method `time' for main (NameError)
                    
                    def foo; time; end; foo
                             ^^^^
                        from -e:1:in `<main>'
                exit 1
ruby-3.4.0-preview1 -e:1:in 'Object#foo': undefined local variable or method 'time' for main (NameError)
                    
                    def foo; time; end; foo
                             ^^^^
                        from -e:1:in '<main>'
                exit 1
ruby-3.4.0-preview2 -e:1:in 'Object#foo': undefined local variable or method 'time' for main (NameError)
                        from -e:1:in '<main>'
                exit 1
```

The Old(3.3) output is:

```ruby
from -e:1:in `<main>'
```